### ISSUE TITLE
Symfony 4 - context - dependency Injection - inject service

### ISSUE THREAD

**Issue by blixit** [Behat/Behat#1180]
Hi, 
I spent 2 days on it and I'm getting crazy. 
I'm trying to inject either kernel or container or any other service into a context. 
However, my service (i.e '@kernel' ) is interpreted as the string '@kernel'. 

I followed the rafal-muszynski tutorial (https://blog.rafalmuszynski.pl/how-to-configure-behat-with-symfony-4/) and I installed behat/symfony2-extension but I got the error:

> PHP Warning:  require_once(/var/www/html/symfony/tracking): failed to open stream: Success in /var/www/html/symfony/tracking/vendor/symfony/dependency-injection/ContainerBuilder.php on line 1099
PHP Fatal error:  require_once(): Failed opening required '/var/www/html/symfony/tracking/' (include_path='.:/usr/share/php') in /var/www/html/symfony/tracking/vendor/symfony/dependency-injection/ContainerBuilder.php on line 1099

:warning: /var/www/html/symfony/tracking is my project directory

I've also noticed there's not any behat recipe for symfony 4.

So, what is the best approach to install behat on symfony 4 ? And how to inject service into context ?
Thanks for help

EDIT: following @timiTao suggestion, here are my files:

```yml
# behat.yml.dist
default:
    suites:
        default:
            contexts:
                - FeatureContext:
                    kernel: '@kernel'
    extensions:
        Behat\Symfony2Extension:
            kernel:
                class: App\Kernel
```
features/demo.feature
```feature
Feature: Demo
  Scenario: This is a scenario
    When I test Behat
    Then It should work
```
features/bootstrap/FeatureContext.php
```php
<?php

namespace App\Behat\Context;

use Behat\Behat\Context\Context;
use Behat\Gherkin\Node\PyStringNode;
use Behat\Gherkin\Node\TableNode;
use Behat\Symfony2Extension\Context\KernelAwareContext;
use Behat\Symfony2Extension\Context\KernelDictionary;
use Symfony\Component\HttpKernel\KernelInterface;

/**
 * Defines application features from the specific context.
 */
class FeatureContext implements KernelAwareContext
{
    use KernelDictionary;
    
    /**
     * Initializes context.
     *
     * Every scenario gets its own context instance.
     * You can also pass arbitrary arguments to the
     * context constructor through behat.yml.
     */
    public function __construct(
        KernelInterface $kernel
    )
    {
        var_dump(get_class($kernel));
    }
}
```

**Comment by spolischook** [comment]
There should be KernelAwareContext I guess. It's not worked out of the box - something should happen for Behat that it aware about Kernel.

**Comment by blixit** [comment]
@spolischook @timiTao 
I just edited my first post with my files.
For version, 
`./vendor/bin/behat -V` gives me the same error I described sooner

but i guess I'm using "behat/behat": "^3.5@dev" (says my composer.json file)

**Comment by blixit** [comment]
I also found that the definition of the app/kernel file is not complete (it's probably a normal behavior), just by opening the current file: 
`./vendor/symfony/dependency-injection/ContainerBuilder.php`
To fix thix error, I tried to tell symfony which file to load in order to create the service. 

```php
namespace Symfony\Component\DependencyInjection;
/**
 * ContainerBuilder is a DI container that provides an API to easily describe services.
 *
 * @author Fabien Potencier <fabien@symfony.com>
 */ 
class ContainerBuilder extends Container implements TaggedContainerInterface {
// ... line 1063
    private function createService(Definition $definition, array &$inlineServices, $isConstructorArgument = false, $id = null, $tryProxy = true)

        // ... line 1098
        var_dump($definition->getFile()); // <= definition file is not complete . I get my project directory

        // my pseudo **and very temporary** fix to avoid error
        if($definition->getClass() == "App\Kernel"){
            $definition->setFile($definition->getFile()."vendor/autoload.php"); // I add path to vendor/autoload.php
        }
        // end of my fix

        // the code that generates the bug
        if (null !== $definition->getFile()) {
            require_once $parameterBag->resolveValue($definition->getFile());
        }
```

But after that, the error disapears and I got a new one

`In UninitializedContextEnvironment.php line 44:
   FeatureContext context class not found and can not be used.`

I really hope you will have a better understanding of what that means

**Comment by blixit** [comment]
I checked the vendor/behat and I followed the 'testapp' then I changed my FeatureContext content

```php
//namespace App\Behat\Context;  // I removed namespace

use Behat\Behat\Context\Context;
use Behat\Gherkin\Node\PyStringNode;
use Behat\Gherkin\Node\TableNode;
use Behat\Symfony2Extension\Context\KernelAwareContext;
use Behat\Symfony2Extension\Context\KernelDictionary;
use Symfony\Component\HttpKernel\KernelInterface;

/**
 * Defines application features from the specific context.
 */
class FeatureContext implements KernelAwareContext
{

    private $kernel;
    private $containerParameters;
    private $parameterKey;

    public function __construct( KernelInterface $kernel)
    {
        $this->setKernel($kernel);
    }

    /**
     * Sets Kernel instance.
     *
     * @param KernelInterface $kernel
     */
    public function setKernel(KernelInterface $kernel)
    {
        $this->kernel = $kernel;
    }
}
```

Now I have this error: 

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to FeatureContext::__construct() must implement interface Symfony\Component\HttpKernel\KernelInterface, string given in /var/www/html/symfony/tracking/features/bootstrap/FeatureContext.php:22
Stack trace:
#0 [internal function]: FeatureContext->__construct('App\\kernel')
#1 /var/www/html/symfony/tracking/vendor/behat/behat/src/Behat/Behat/Context/ContextFactory.php(133): ReflectionClass->newInstanceArgs(Array)
#2 /var/www/html/symfony/tracking/vendor/behat/behat/src/Behat/Behat/Context/ContextFactory.php(88): Behat\Behat\Context\ContextFactory->createInstance(Object(ReflectionClass), Array)
#3 /var/www/html/symfony/tracking/vendor/behat/behat/src/Behat/Behat/Context/Environment/Handler/ContextEnvironmentHandler.php(121): Behat\Behat\Context\ContextFactory->createContext('FeatureContext', Array, Array)
#4 /var/www/html/symfony/tracking/vendor/behat/behat/src/Behat/Testwork/Environment/EnvironmentManager.php(93): Behat\Behat\Context\Environment\Handler\ContextEnvironmentHan in /var/www/html/symfony/tracking/features/bootstrap/FeatureContext.php on line 22
```

**Comment by pueppiblue** [comment]
Hi,
we ran in the same Fatal Error you formerly received as well:

> Warning: require_once(/var/www/html): failed to open stream: Success in /var/www/html/vendor/symfony/dependency-injection/ContainerBuilder.php on line 1089
> 
> Fatal error: require_once(): Failed opening required '/var/www/html/' (include_path='.:/usr/local/lib/php') in /var/www/html/vendor/symfony/dependency-injection/ContainerBuilder.php on line 1089
> 

Turned out we used the behat/symfony2-extension with dev-master stability.
"Downgrading" to 2.1.5 resolved the issue.

Regarding your parameter issue:
Maybe it's because you are missing the config entry to point the symfony2-extension to the bootstrap file, which basically only loads the .env file.
```yaml
 extensions:
         Behat\Symfony2Extension:
             kernel:
                 bootstrap: features/bootstrap/bootstrap.php
                 class: App\Kernel
```

**Comment by blixit** [comment]
ok. It looks like the problem is solved by explicitly installing 2.1.5 and without having to add                  
bootstrap: features/bootstrap/bootstrap.php

**Comment by nickreynke** [comment]
Thank you @blixit!

**Comment by nickreynke** [comment]
Thank you @blixit and @pueppiblue!

**Comment by jcuerdo** [comment]
> Hi,
> we ran in the same Fatal Error you formerly received as well:
> 
> > Warning: require_once(/var/www/html): failed to open stream: Success in /var/www/html/vendor/symfony/dependency-injection/ContainerBuilder.php on line 1089
> > Fatal error: require_once(): Failed opening required '/var/www/html/' (include_path='.:/usr/local/lib/php') in /var/www/html/vendor/symfony/dependency-injection/ContainerBuilder.php on line 1089
> 
> Turned out we used the behat/symfony2-extension with dev-master stability.
> "Downgrading" to 2.1.5 resolved the issue.
> 
> Regarding your parameter issue:
> Maybe it's because you are missing the config entry to point the symfony2-extension to the bootstrap file, which basically only loads the .env file.
> 
> ```yaml
>  extensions:
>          Behat\Symfony2Extension:
>              kernel:
>                  bootstrap: features/bootstrap/bootstrap.php
>                  class: App\Kernel
> ```

Fixed for me  downgrading to 2.1.5 resolved the issue. Thanks
